File: sdks/python/examples/sidecar_convert.py
#!/usr/bin/env python3
"""Sidecar mode: openrate as a child process on 127.0.0.1, spoken to over HTTP.

    python3 examples/sidecar_convert.py                       # managed child
    python3 examples/sidecar_convert.py https://openrate.io    # somebody else's

You do not run a server. `openrate.Client()` finds the binary
($OPENRATE_BINARY, the bundled openrate/bin/openrate, or `openrate` on PATH),
picks a free loopback port, launches it, waits for /readyz, and terminates it at
exit. Given a URL instead, it spawns nothing and talks to the server you already
run — the same client either way, because it is the same API.

UNLIKE DIRECT MODE, THIS FETCHES. A server refreshes on startup and on its
interval; that is what a server is for. If you want a process that provably
sends no packets, that is `examples/direct_convert.py` — an engine with no
refresher.

This example needs network access for the managed case, because a freshly
started openrate has an empty book until its first fetch lands. Waiting for
/readyz rather than /healthz is what makes that a named failure ("ecb:
connection refused") instead of every conversion coming back as an unknown
currency pair.

    OPENRATE_READY_TIMEOUT   seconds to wait for the first rate (default 30)
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import openrate  # noqa: E402
from openrate import Client, OpenRateSidecarError  # noqa: E402

SOURCES = "ecb"  # one fast source keeps the example quick; the default is four
READY_TIMEOUT = float(os.environ.get("OPENRATE_READY_TIMEOUT", "30"))


def main(argv: list[str]) -> int:
    url = argv[0] if argv else None

    # try/finally, not just atexit: an exception below must not leave a server
    # process behind for the rest of this interpreter's life. `with` does the
    # same thing, and closes nothing when the client points at somebody else's
    # server — it did not start it.
    if url is None:
        # start() returns when /readyz says a conversion will succeed, so there
        # is no separate "wait for the book" step here and no window in which
        # this example could convert against an empty one. If no rate arrives it
        # raises, naming the source that failed.
        openrate.start(sources=SOURCES, base_currency="ZAR", timeout=READY_TIMEOUT)

    with Client(url) as client:
        print(f"base url:     {client.base_url}")
        print(f"managed:      {'no — pointing at an existing server' if url else 'yes'}")
        print(f"healthy:      {client.healthy()}   (liveness: the process is up)")
        print(f"ready:        {client.ready()}   (readiness: it has rates)\n")

        if url is not None:
            # Somebody else's server can be up and still empty; we did not wait
            # for it at startup because we did not start it.
            client.wait_ready(READY_TIMEOUT)

        meta = client.meta()
        print(f"meta          default base {meta['default_base']}, "
              f"{len(meta['currencies'])} currencies")
        for source in meta["sources"]:
            print(f"              source {source['name']}: {source['edges']} edges, "
                  f"last_error={source.get('last_error') or 'none'}")

        answer = client.convert("USD", "ZAR", 100)
        rate = answer["rate"]
        print(f"convert       {answer['amount']} {answer['from']} = "
              f"{answer['result']:.2f} {answer['to']}")
        print(f"              rate {rate['rate']:.4f}, {rate['hops']} hop(s), "
              f"path {' -> '.join(rate['path'])}, sources {rate['sources']}")
        print(f"              quality {rate['quality']['grade']}, "
              f"{rate['age_sec']:.0f}s old")

        book = client.rates("ZAR")
        sample = sorted(book["rates"])[:6]
        print(f"rates         {len(book['rates'])} currencies against ZAR; "
              f"first few: {', '.join(sample)}")

        # The one deliberate difference between the two modes: an unknown base
        # answers 200 with an empty book here, where Engine.rates() raises.
        empty = client.rates("XXX")
        print(f"unknown base  HTTP 200 with {len(empty['rates'])} rates "
              f"(direct mode raises instead — the one deliberate difference)")

        # An unknown pair is a 404 with a JSON body, where direct mode raises
        # OpenRateError carrying plain text. Same failure, two shapes.
        try:
            client.convert("USD", "XXX", 1)
        except OpenRateSidecarError as exc:
            print(f"error         {exc}")

    print("\nsidecar stopped" if url is None else "\nleft the remote server alone")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main(sys.argv[1:]))
    except OpenRateSidecarError as exc:
        print(f"openrate: {exc}", file=sys.stderr)
        openrate.stop()
        sys.exit(1)

File: sdks/elixir/examples/sidecar_convert.exs
# SIDECAR — the SDK spawns the `openrate` binary on a loopback port through an
# Erlang Port, waits for /healthz, and reaps it when the GenServer stops. You
# never run a server by hand.
#
#   cd sdks/elixir && mix run examples/sidecar_convert.exs
#
# Environment:
#   OPENRATE_BINARY   path to the openrate binary (default: bundled priv/bin, then PATH)
#   OPENRATE_SOURCES  comma-separated FX sources (default here: ecb)
#
# For Elixir this is not the "fallback" mode — it is the mode. README.md sets
# out why an in-process NIF is the wrong trade for this library.

sources = System.get_env("OPENRATE_SOURCES", "ecb")

{:ok, base} =
  case Openrate.start(sources: sources, ui: false) do
    {:ok, base} ->
      {:ok, base}

    {:error, reason} ->
      IO.puts(:stderr, "could not start the sidecar: #{inspect(reason)}")
      System.halt(1)
  end

IO.puts("sidecar : #{base}")

# try/after, so a failure below still stops the child rather than leaving an
# orphaned server holding a port. Openrate.Sidecar would also reap it at VM
# shutdown; being explicit frees the port the moment we are done.
try do
  # The server answers /healthz before its first fetch completes, so wait for
  # the book rather than racing it.
  deadline = System.monotonic_time(:millisecond) + 60_000

  meta =
    Stream.repeatedly(fn ->
      Process.sleep(250)
      Openrate.meta()
    end)
    |> Stream.take_while(fn _ -> System.monotonic_time(:millisecond) < deadline end)
    |> Enum.find({:error, :timeout}, fn
      {:ok, %{"currencies" => [_ | _]}} -> true
      _ -> false
    end)

  {:ok, meta} = meta

  IO.puts(
    "meta    : base #{meta["default_base"]}, #{length(meta["currencies"])} currencies " <>
      "from #{length(meta["sources"])} source(s)"
  )

  Enum.each(meta["sources"], fn source ->
    # last_error is absent, not nil, when a source succeeded.
    detail =
      case Map.get(source, "last_error") do
        nil -> "#{source["edges"]} edges"
        "" -> "#{source["edges"]} edges"
        err -> "error: #{err}"
      end

    IO.puts("source  : #{String.pad_trailing(source["name"], 10)} #{detail}")
  end)

  # 1. Convert — the same JSON document the C ABI's "convert" method returns.
  {:ok, r} = Openrate.convert("EUR", "USD", 100)
  rate = r["rate"]

  IO.puts(
    "convert : #{r["amount"]} #{r["from"]} = #{Float.round(r["result"] / 1, 4)} #{r["to"]} " <>
      "(rate #{rate["rate"]}, #{rate["hops"]} hop, #{Enum.join(rate["sources"], ",")})"
  )

  # 2. A cross-rate the sources do not publish directly.
  {:ok, r} = Openrate.convert("EUR", "ZAR", 100)

  IO.puts(
    "crossed : 100 EUR = #{Float.round(r["result"] / 1, 2)} ZAR " <>
      "via #{Enum.join(r["rate"]["path"], "→")}"
  )

  # 3. The all-pairs snapshot.
  {:ok, rates} = Openrate.rates("ZAR")
  sample = rates["rates"] |> Map.keys() |> Enum.sort() |> Enum.take(5) |> Enum.join(", ")
  IO.puts("rates   : base #{rates["base"]}, #{map_size(rates["rates"])} currencies (#{sample}, …)")

  # 4. The error path. Over HTTP an error is a status code and a JSON body,
  #    where the C ABI hands back a plain string in *err.
  {:error, {:http, status, message}} = Openrate.convert("XXX", "YYY", 1)
  IO.puts("error   : HTTP #{status} #{message}")

  # 5. Crash isolation — the property a NIF would spend. A process that dies
  #    mid-call takes nothing with it, and the sidecar is untouched.
  {:ok, agent} = Agent.start(fn -> nil end)
  Process.unlink(agent)
  ref = Process.monitor(agent)
  Process.exit(agent, :kill)

  receive do
    {:DOWN, ^ref, :process, _, reason} ->
      IO.puts("isolate : a worker died (#{inspect(reason)}); the VM and the sidecar are fine")
  end

  {:ok, _} = Openrate.meta()
  IO.puts("          openrate still answering after that")
after
  Openrate.stop()
end

IO.puts("stopped : ok")
